WIP: Fixed wing control API

RomanBapst opened the pull request:
### Solved Problem
Refactor fixed wing position controller to create flexible control API according to diagram provided.

![Screenshot from 2024-11-29 10-10-48](https://github.com/user-attachments/assets/974e881a-1148-4345-baf2-34998f618b5b)


**TODO**:

- [ ] split controllers into separate modules where it makes sense
- [ ] refactor longitudinal control, only lateral part done until now
- [ ] handle constraints (throttle, roll) -> similar to what we do for multicopters


### Changelog Entry
For release notes:
```
Feature/Bugfix XYZ
New parameter: XYZ_Z
Documentation: Need to clarify page ... / done, read docs.px4.io/...
```

### Alternatives
We could also ...

### Test coverage
- Unit/integration test: ...
- Simulation/hardware testing logs: https://review.px4.io/

### Context
Related links, screenshot before/after, video

KonradRudin commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,680 @@
+		fw_longitudinal_control_setpoint_s longitudinal_sp = {empty_longitudinal_control_setpoint};
+		_fw_longitudinal_ctrl_sub.copy(&longitudinal_sp);
+
+		if (longitudinal_sp.timestamp > 0 && hrt_elapsed_time(&longitudinal_sp.timestamp) < 1_s) {
```
I think that would need to be set better, Controller need to know if it is in charge or not. Like with the flags from the commander.

KonradRudin reviewed (commented):
I like the restructuring, having the fixed wing position control more separated into the mode and the actual control. As well as having more intermediate options. 
I went through all of it. but not in the complete detail yet, just when i saw something i mentioned it. 
Biggest issue for me is that we don't yet have a clear way to tell, that the longitudinal and lateral controller should be enabled or not. Should be in the vehicle control mode flags propably like _vcontrol_mode.flag_control_rates_enabled

RomanBapst commented on src/lib/npfg/npfg.hpp:
```diff
@@ -1,616 +0,0 @@
-	 * @return estimate of certainty of the correct functionality of the npfg roll setpoint in [0, 1]. Can be used to define proper mitigation actions.
-	 */
-
-	float canRun(const vehicle_local_position_s &local_pos, bool is_wind_valid) const;
```
@KonradRudin It's now done in `getGuidanceQualityFactor` in the FwLatLon module

Jaeyoung-Lim reviewed (commented):
Sorry that I am late to comment on this PR. I think there would be a lot of value of creating a FW API.

The api proposed have separate messages for longitudinal and lateral gudiance. This seems to be built on the assumption that lateral and longitudinal control (TECS) would be controlled in a decoupled way. 

However, it would have been nicer if the api would be designed for more "generic" fixed-wing tasks, rather than exposing configurations of the underlying interface. IMO, an api should fully define what the vehicle "should" do for a given task (e.g. path following), rather than expose the underlying implementation.

- Lateral/longitudinal controller decoupling is a design choice for a specific implementation. This would make it hard for other controllers to share the API, if the controllers do not have such decoupling (e.g. 3D NDI acceleration-based control).

- The decoupling is not only not necessary for the task of path following, but can even limit what kind of paths we can follow in 3D. For example, if you try to track an inclined loiter(or any complex 3D path of arbitrary shape) the height rate setpoint in the `FwLongitudinalSetpoint.msg` is no longer usable, since the information on the 3D geometry is already lost (e.g. height rate depends on the speed and direction of the vehicle to stay on the path).

Is there a good reason that we want to enforce the separation of lateral/longitudinal control on the API side?
@tstastny

RomanBapst commented:
@Jaeyoung-Lim Thanks for your input on this!

> The api proposed have separate messages for longitudinal and lateral gudiance. This seems to be built on the assumption that lateral and longitudinal control (TECS) would be controlled in a decoupled way.

Yes, and so far lateral and longitudinal control have almost completely been decoupled in PX4. The only coupling I am aware of is that NPFG used to be able to change the airspeed setpoint. But other than that, it was decoupled.
So far this architecture has brought us a long way and I don't think it's terrible.

> However, it would have been nicer if the api would be designed for more "generic" fixed-wing tasks, rather than exposing configurations of the underlying interface. IMO, an api should fully define what the vehicle "should" do for a given task (e.g. path following), rather than expose the underlying implementation.

You are not limited to the exposed interfaces presented in this PR. If you want to have your fancy NDI controller running offboard and send attitude / rate or even actuator commands directly, you are free to do so, those interfaces I think are already there.

> Lateral/longitudinal controller decoupling is a design choice for a specific implementation. This would make it hard for other controllers to share the API, if the controllers do not have such decoupling (e.g. 3D NDI acceleration-based control).

I think it's a design choice which solves 95% of the use case that people would want. I think that's pretty good.

> The decoupling is not only not necessary for the task of path following, but can even limit what kind of paths we can follow in 3D. For example, if you try to track an inclined loiter(or any complex 3D path of arbitrary shape) the height rate setpoint in the FwLongitudinalSetpoint.msg is no longer usable, since the information on the 3D geometry is already lost (e.g. height rate depends on the speed and direction of the vehicle to stay on the path).

Could you please specify the **exact interface** that you would need to solve whatever use case you are trying to solve?
If it's a valid use case and it cannot be done with the current API, then we could always add it.


Generally, I have a feeling that your expectation is that ONE uorb topic defines exactly ONE control API. I don't think this needs to be true necessarily. I think you will use some kind of SDK, which will expose one interface to the developer, but the SDK itself can use multiple uorb topics.

tstastny commented:
Hi. Thanks @RomanBapst and @KonradRudin for putting in work to finally start cleaning up the the fixed-wing high-level control infrastructure a bit! I haven't had time to fully go through the details, but I am a little confused what specifically is the "API" here. But there may just be some difference in everyone's semantics...

The uorb topics in the figure look more like **logging topics** to me, as I find it odd an application would send e.g. pitch setpoint, throttle setpoint, altitude setpoint, and height rate setpoint all out into the ether at the same time? Maybe @RomanBapst you can de-confuse me on the strategy? :) Are there multiple uorbs being bundled and sent at once from e.g. the ROS2 mode? Are these messages just defining what each block (controller) needs as input? And the actual exposed API will be something more basic, with each control block parsing it to convert to the right inputs to each?

On the topic of alternate APIs (@Jaeyoung-Lim ), the block diagram itself looks as though it's not terribly far away from supporting a wide array of things. e.g. if there were a single 3d instantaneous path setpoint API, the height component of it could still be routed to Alt-Ctrl, the vertical incline of the path then converted to a height rate setpoint that could be used as feedforward, added to the feedback height rate from Alt-Ctrl, and then sent off to TECS (or Velocity NDI, or whatever). It looks as though that is actually already somewhat happening in the diagram?

So long as the blocks on the diagram are decoupled as shown, and new urobs can be made with a parser to spit out the resp. input msgs each controller one is using needs, I would imagine it's possible to do, right? Or maybe I miss something on the routing details? (I am a bit foggy on things now after not looking for a long time)

RomanBapst commented:
> The uorb topics in the figure look more like logging topics to me, as I find it odd an application would send e.g. pitch setpoint, throttle setpoint, altitude setpoint, and height rate setpoint all out into the ether at the same time? Maybe @RomanBapst you can de-confuse me on the strategy? :) 

@tstastny I have changed the title of the PR to not contain the word API, as it seems to be confusing.
You can call the new uorb messages an API, but it's definitely not the API we want to expose 1 to 1 to the end developer.
The idea would be that the SDK which is used for development (auterion SDK or whatever SDK you like), converts the developer facing controller setpoint into the uorb messages defined in this PR.

Also note, that the current uorb messages contain certain fields I wish would not be there, but since it was a priority to not break existing functionality with this refactor, I added them regardless. I am all in favor to remove fields like  `heading_sp_runway_takeoff` and `pitch_sp` from the message.
It's the fixed wing takeoff and the fixed wing landing modes which requires these fields.

I'm also adding a diagram here which I hope explains a bit better the architecture in terms of modularity.
![image](https://github.com/user-attachments/assets/b857271f-a6a2-431b-86e1-0074d237d74a)

RomanBapst commented on msg/FwLongitudinalControlSetpoint.msg:
```diff
@@ -0,0 +1,11 @@
+float32 height_rate_setpoint # NAN if not controlled directly, used as feeforward if altitude_setpoint is finite [m/s]
+float32 altitude_setpoint  # NAN if not controlled, MSL [m]
+float32 equivalent_airspeed_setpoint # [m/s]
+float32 pitch_sp # NAN if not controlled, overrides total energy controller [rad]
```
@KonradRudin The reason that they are there is because the fixed wing takeoff and landing modes do make use of setting pitch / roll setpoints directly. I thought about having those modes publish attitude setpoints directly, however, the issue is that the FwLatLonControl module does not know about this and keeps on publishing attitude setpoints as well. The correct approach would be to set the control flags for attitude and rates only, but that requires those modes to be separated in commander. I don't yet have a good solution for this.

RomanBapst commented on src/modules/fw_pos_control/runway_takeoff/RunwayTakeoff.h:
```diff
@@ -116,13 +116,13 @@ class __EXPORT RunwayTakeoff : public ModuleParams
+	float getPitch();
 
 	/**
 	 * @param external_roll_setpoint Externally commanded roll angle setpoint (usually from path navigation) [rad]
```
@KonradRudin I don't understand what you mean here.

sfuhrer commented on msg/FwLateralControlSetpoint.msg:
```diff
@@ -0,0 +1,12 @@
+float32 lateral_acceleration_setpoint # NAN if not controlled directly, used as feedforward if either course setpoint or airspeed_reference_direction is finite, [m/s^2]
+float32 roll_sp # TODO: remove, only for testing
+
+float32 heading_sp_runway_takeoff # [-pi, pi] heading setpoint for runway takeoff
```
removed functionality. Can think about adding it again but through separate messages.

RomanBapst commented on src/modules/fw_pos_control/FixedwingPositionControl.cpp:
```diff
@@ -930,90 +724,125 @@ FixedwingPositionControl::control_auto(const float control_interval, const Vecto
+	longitudinal_control_limits.throttle_max = (_landed
+			|| !_local_pos.v_z_valid) ? _param_fw_thr_min.get() : _param_fw_thr_max.get();
+
+	_longitudinal_ctrl_limits_pub.publish(longitudinal_control_limits);
```
@KonradRudin I worked on this and now the limits are only published when they change.

perrrre commented on msg/FwLateralControlSetpoint.msg:
```diff
@@ -0,0 +1,10 @@
+
+# NOTE: At least one of course_setpoint, airspeed_reference_direction, or lateral_acceleration_setpoint must be finite
+
+float32 course_setpoint 		# [-pi, pi], NAN if not controlled directly
+float32 airspeed_reference_direction    # [-pi, pi], angle of desired airspeed vector, NAN if not controlled directly, used as feedforward if course setpoint is finite
```
Looks like the other elements in the messages are described with units, [rad] could be added here as well.

sfuhrer replied:
and we need to add if it's a true airspeed or calibrated/indicated/equivalent

perrrre commented on msg/FwLateralControlSetpoint.msg:
```diff
@@ -0,0 +1,10 @@
+float32 course_setpoint 		# [-pi, pi], NAN if not controlled directly
+float32 airspeed_reference_direction    # [-pi, pi], angle of desired airspeed vector, NAN if not controlled directly, used as feedforward if course setpoint is finite
+float32 lateral_acceleration_setpoint 	# [m/s^2], NAN if not controlled directly, used as feedforward if either course setpoint or airspeed_reference_direction is finite
+float32 roll_sp 			# TODO: remove, only for testing
```
remove

sfuhrer commented on msg/LongitudinalControlLimits.msg:
```diff
@@ -0,0 +1,15 @@
+float32 equivalent_airspeed_min # [m/s]
+float32 equivalent_airspeed_max # [m/s]
+
+float32 speed_weight # [0,2], 0=pitch controls altitude only, 2=pitch controls airspeed only
+
+bool enforce_low_height_condition # if true, total energy controller will use lower altitude control time constant
+bool disable_underspeed_protection
```
@KonradRudin are you referring to `disable_underspeed_protection`? We need to somehow pass to TECS to not do underspeed detection during takeoff and landing - or even better, handle it outside of TECS.

perrrre commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,764 @@
+	return can_run_factor * (lateral_accel_sp);
+}
+float FwLateralLongitudinalControl::mapLateralAccelerationToRollAngle(float lateral_acceleration_sp) const {
+	return  atanf(lateral_acceleration_sp * 1.0f / CONSTANTS_ONE_G);
```
After discussion we can remove the 1.0f from the function.

perrrre commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,764 @@
+				roll_sp = mapLateralAccelerationToRollAngle(lateral_accel_sp);
+			}
+
+			fw_lateral_control_setpoint_s status = {
+				.timestamp = _lat_control_sp.timestamp,
+				.course_setpoint = _lat_control_sp.course_setpoint,
+				.airspeed_reference_direction = airspeed_reference_direction,
+				.lateral_acceleration_setpoint = lateral_accel_sp,
+				.roll_sp = roll_sp // TODO: just for logging, can be removed later
+			};
```
Add a comment that its only for logging and not consumed by any other

sfuhrer replied:
done

perrrre commented on src/lib/npfg/AirspeedReferenceController.cpp:
```diff
@@ -0,0 +1,63 @@
+/****************************************************************************
+ *
+ * Copyright (c) 2024 PX4 Development Team. All rights reserved.
```
```suggestion
 * Copyright (c) 2025 PX4 Development Team. All rights reserved.
```
saw it in a few other places as well

sfuhrer replied:
done

perrrre commented on src/lib/npfg/CourseToAirspeedRefMapper.hpp:
```diff
@@ -0,0 +1,138 @@
+	 * @param[in] wind_cross_bearing 2D cross product of wind velocity and bearing vector [m/s]
+	 * @param[in] wind_dot_bearing 2D dot product of wind velocity and bearing vector [m/s]
+	 * @param[in] wind_speed Wind speed [m/s]
+	 * @param[in] min_ground_speed Minimum commanded forward ground speed [m/s]
```
min_ground_speed is no longer an input

perrrre commented on src/lib/npfg/AirspeedReferenceController.cpp:
```diff
@@ -0,0 +1,63 @@
+	} else {
+		// airspeed/airspeed_ref is used to scale any incremented airspeed reference back to the current airspeed
+		// for acceleration commands in a "feedback" sense (i.e. at the current vehicle airspeed)
+		return p_gain_ * cross_air_vel_err;
```
should it be scaled by airspeed_ref?

perrrre replied:
It used to be a an airspeed error, now the error is on the heading. The cross product used to produce [m^2/s^2] now its [m/s], so scaling is not needed. However the comment could be updated.

perrrre commented on src/lib/npfg/npfg.cpp:
```diff
@@ -1,508 +0,0 @@
-	// must be called before trackErrorBound() as it updates time_const_
-	adapted_period_ = adaptPeriod(ground_speed, airspeed, wind_speed, track_error,
-				      path_curvature, wind_vel, unit_path_tangent, feas_on_track_);
-	p_gain_ = pGain(adapted_period_, damping_);
```
Is an adaptive gain not needed anymore?

perrrre commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,764 @@
+			float throttle_sp{NAN};
+
+			if (_fw_longitudinal_ctrl_sub.updated()) {
+				_fw_longitudinal_ctrl_sub.copy(&_long_control_sp);
```
```suggestion
                                fw_longitudinal_control_setpoint_s long_control_sp
				_fw_longitudinal_ctrl_sub.copy(&long_control_sp);
```
Looks like it's only in this scope so definition can be within the scope instead of the headerfile

sfuhrer replied:
yeah but `_fw_longitudinal_ctrl_sub.updated()` may not always be true, and in these cases we still want to execute the code here but with the value form a previous update.

perrrre commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,764 @@
+}
+
+void FwLateralLongitudinalControl::setDefaultLongitudinalControlLimits() {
+		_long_limits.timestamp = hrt_absolute_time();
+		_long_limits.equivalent_airspeed_min = _performance_model.getMinimumCalibratedAirspeed();
+		_long_limits.equivalent_airspeed_max = _performance_model.getMaximumCalibratedAirspeed();
+		_long_limits.pitch_min = radians(_param_fw_p_lim_min.get());
+		_long_limits.pitch_max = radians(_param_fw_p_lim_max.get());
+		_long_limits.throttle_min = _param_fw_thr_min.get();
+		_long_limits.throttle_max = _param_fw_thr_max.get();
+		_long_limits.climb_rate_target = _param_climbrate_target.get();
+		_long_limits.sink_rate_target = _param_sinkrate_target.get();
+		_long_limits.disable_underspeed_protection = false;
+		_long_limits.enforce_low_height_condition = false;
```
```suggestion
		_long_limits.timestamp = hrt_absolute_time();
		_long_limits.equivalent_airspeed_min = _performance_model.getMinimumCalibratedAirspeed();
		_long_limits.equivalent_airspeed_max = _performance_model.getMaximumCalibratedAirspeed();
		_long_limits.pitch_min = radians(_param_fw_p_lim_min.get());
		_long_limits.pitch_max = radians(_param_fw_p_lim_max.get());
		_long_limits.throttle_min = _param_fw_thr_min.get();
		_long_limits.throttle_max = _param_fw_thr_max.get();
		_long_limits.climb_rate_target = _param_climbrate_target.get();
		_long_limits.sink_rate_target = _param_sinkrate_target.get();
		_long_limits.disable_underspeed_protection = false;
		_long_limits.enforce_low_height_condition = false;
```
The rows has 2 indents

perrrre commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,764 @@
+						   _long_limits.disable_underspeed_protection,
+						   _long_control_sp.height_rate_setpoint
+						  );
+			pitch_sp = PX4_ISFINITE(_long_control_sp.pitch_sp) ? _long_control_sp.pitch_sp : _tecs.get_pitch_setpoint();
```
The pitch setpoint used to have a protection in case the tecs is not running:
`
	if (_tecs_is_running) {
		return _tecs.get_pitch_setpoint() + radians(_param_fw_psp_off.get());
	}

	// return level flight pitch offset to prevent stale tecs state when it's not running
	return radians(_param_fw_psp_off.get());
`

Shall we add a method for that here?

sfuhrer replied:
I would say no, there is no reason TECS is not running at this point. And if there is a condition where it's not running we need to fix it.
But good that you highlighted it, made me clean up some other parts.

perrrre commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,764 @@
+						   _long_control_sp.height_rate_setpoint
+						  );
+			pitch_sp = PX4_ISFINITE(_long_control_sp.pitch_sp) ? _long_control_sp.pitch_sp : _tecs.get_pitch_setpoint();
+			throttle_sp = PX4_ISFINITE(_long_control_sp.thrust_sp) ? _long_control_sp.thrust_sp : _tecs.get_throttle_setpoint();
```
The throttle also had some protection:

`
	if (_tecs_is_running) {
		return min(_tecs.get_throttle_setpoint(), 1.f);
	if (_control_mode.flag_control_position_enabled) {
	}

	// return 0 to prevent stale tecs state when it's not running
	return 0.0f;`

sfuhrer commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,764 @@
+			_longitudinal_ctrl_status_pub.publish(longitudinal_control_status);
+
+			float roll_sp {NAN};
+			float yaw_sp {NAN};
```
done

sfuhrer commented:
Rebased on main.

mahima-yoga commented on src/lib/npfg/AirspeedReferenceController.cpp:
```diff
@@ -0,0 +1,64 @@
+	const Vector2f airspeed_vector = Vector2f{cosf(heading), sinf(heading)} * airspeed;
+	const Vector2f airspeed_sp_vector_unit = Vector2f{cosf(heading_sp), sinf(heading_sp)};
+
+	const float dot_air_vel_err = airspeed_vector.dot(airspeed_sp_vector_unit);
```
@RomanBapst @sfuhrer 

Should this be 
`const float dot_air_vel_err = 1.f - airspeed_vector.dot(airspeed_sp_vector_unit)` ? 

I think with the current implementation the error is increasing as the two vectors become more parallel

sfuhrer replied:
The same is already in the [current implementation](https://github.com/PX4/PX4-Autopilot/blob/55090618038537b493c049af82d052ce485d1c82/src/lib/npfg/npfg.cpp#L481). Which ofc doesn't imply that it has to be correct.
But it's just to check the angle between the vectors is higher than 90° or not, and for the current check looks correct to me.

tstastny commented:
regarding https://github.com/PX4/PX4-Autopilot/pull/24629#issuecomment-2783070840

> @tstastny , @mahima-yoga has added the handling of infeasible bearing in [this commit](https://github.com/PX4/PX4-Autopilot/pull/24629/commits/63e1b08d11a0dab4813ad6b673ad290a80c53e8f),in case you want to check it out.
> 
> Compared to the previous implementation we've removed some options and only kept the default case (`NPFG_EN_MIN_GSP=1`, `NPFG_WIND_REG=1`, `NPFG_TRACK_KEEP=1`).
> 
> There's now also a bunch of unit tests that test and showcase the handling of excess wind with the current implementation.

@sfuhrer the minimum groundspeed logic is currently not used in the refactor, right? i had discussed with @RomanBapst that this airspeed setpoint handling was better decoupled from the main guidance logic. but could still be used as a separate parallel guidance layer in front of the airspeed controller (TECS) if you want these features still. otherwise yes squashing these params is probably fine as default.

sfuhrer commented:
> @sfuhrer the minimum groundspeed logic is currently not used in the refactor, right? i had discussed with @RomanBapst that this airspeed setpoint handling was better decoupled from the main guidance logic. but could still be used as a separate parallel guidance layer in front of the airspeed controller (TECS) if you want these features still. otherwise yes squashing these params is probably fine as default.

@tstastny yes it is used, see [this function](https://github.com/PX4/PX4-Autopilot/blob/46933ab1193c9c74c3f15429966b8c7a3acdcafd/src/lib/npfg/CourseToAirspeedRefMapper.cpp#L73) and the [function's unit tests here](https://github.com/PX4/PX4-Autopilot/blob/46933ab1193c9c74c3f15429966b8c7a3acdcafd/src/lib/npfg/NpfgTest.cpp#L151). It is now indeed de-coupled form the guidance (which outputs the course and lateral_acceleration only, but doesn't do the airspeed augmentation.

sfuhrer commented:
> https://github.com/PX4/PX4-Autopilot/pull/24056#issuecomment-2842293280

@hamishwillee can you help out here? The Docs CI complains about not finding a param that I added in this very PR - do we need to manually update the parameter reference to fix that?

RomanBapst commented on src/modules/fw_mode_manager/ControllerConfigurationHandler.hpp:
```diff
@@ -0,0 +1,113 @@
+			diff = PX4_ISFINITE(current_value) ? (fabsf(new_value - current_value) > FLT_EPSILON) : true;
+
+		} else {
+			diff = PX4_ISFINITE(current_value);
```
@KonradRudin If the new value is not finite, it means that the new value is default. So a change should be published if the old value is finite and thus not the default.

sfuhrer replied:
Yes looks correct to me.

RomanBapst commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,841 @@
+						   airspeed_vector.norm());
+			}
+
+			if (PX4_ISFINITE(_lat_control_sp.lateral_acceleration)) {
```
@KonradRudin That's a good question. I think it would make sense to set the local variable `lateral_accel_sp` to 0, instead of NAN.

sfuhrer replied:
Okay set it to 0.

sfuhrer replied:
717da9171154686a83af06693561965591d22297

RomanBapst commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,841 @@
+			// If the both altitude and height rate are set, set altitude setpoint to NAN
+			const float altitude_sp = PX4_ISFINITE(_long_control_sp.height_rate) ? NAN : _long_control_sp.altitude;
+
+			tecs_update_pitch_throttle(control_interval, altitude_sp,
```
@KonradRudin Airspeed is validated in `adapt_airspeed_setpoint`, I can change the name to `validate_and_adapt_airspeed_setpoint`.

If you set both altitude and height rate setpoin to NAN, then you should get to this statement inside TECS, so should be fine in theory. This would mean that TECS would basically try to control height rate to 0

`if (!PX4_ISFINITE(setpoint.alt) && !PX4_ISFINITE(setpoint.alt_rate)) {
		// neither altitude nor altitude rate is set - reset to current altitude
		_velocity_control_traj_generator.reset(0.f, 0, current_alt);
		altitude_setpoint = current_alt;`

RomanBapst commented on msg/versioned/FixedWingLongitudinalSetpoint.msg:
```diff
@@ -0,0 +1,13 @@
+# Fixed Wing Longitudinal Setpoint message
+# Used by the fw_lateral_longitudinal_control module
+# If not both pitch_direct and throttle_direct are finite, then either altitude or height_rate must be finite.
```
@KonradRudin Hm, now that I have checked what happens when both alt and alt rate setpoint or non-finite, I wonder if we should not change the sentence completely.

mahima-yoga replied:
@RomanBapst Something like: " If pitch_direct and throttle_direct are not both finite, then the controller relies on either altitude or height_rate to control vertical motion. If both altitude and height_rate are NAN, the controller maintains the current altitude." ?

hamishwillee commented:
FYI the flaw bot post above in https://github.com/PX4/PX4-Autopilot/pull/24056#issuecomment-2876837999 is invalid - it is happening because we don't yet update the parameters since moving docs in-tree. Ramon is nearly there.

@RomanBapst Are there any docs to be done out of this? My understanding is existing flight behaviour etc is unchanged but there might be some developer docs about the new APIs? Perhaps also new APIs in the [ROS 2 interface library](https://docs.px4.io/main/en/ros2/px4_ros2_interface_lib.html)?

sfuhrer commented:
> FYI the flaw bot post above in [#24056 (comment)](https://github.com/PX4/PX4-Autopilot/pull/24056#issuecomment-2876837999) is invalid - it is happening because we don't yet update the parameters since moving docs in-tree. Ramon is nearly there.

Okay no worries then for now.

> @RomanBapst Are there any docs to be done out of this? My understanding is existing flight behaviour etc is unchanged but there might be some developer docs about the new APIs? Perhaps also new APIs in the [ROS 2 interface library](https://docs.px4.io/main/en/ros2/px4_ros2_interface_lib.html)?

Yes the existing flight behavior is largely unchanged and I've adapted the corresponding sections in the docs that have changed.

@mahima-yoga is handling the documentation of the interface lib, see https://github.com/PX4/PX4-Autopilot/pull/24785

mahima-yoga commented on msg/versioned/FixedWingLongitudinalSetpoint.msg:
```diff
@@ -0,0 +1,13 @@
+uint64 timestamp                        # time since system start (microseconds)
+
+float32 altitude  			# [m] Altitude setpoint AMSL, not controlled directly if NAN or if height_rate is finite
+float32 height_rate 			# [m/s] NAN if not controlled directly
```
>Need to clarify here if positive value is for up or down? so NED or ENU. Also i like to have an even minimal description instead of only decribing NAN.

addressed

sfuhrer replied:
>Also for future it should be possible to set this as feedforward i guess? How will it work with versioning if only the description changes?

I find that then fitting better in a path setpoint than the one here.

mahima-yoga commented on msg/versioned/FixedWingLateralSetpoint.msg:
```diff
@@ -0,0 +1,11 @@
+uint64 timestamp                        # time since system start (microseconds)
+
+float32 course 				# [rad] [-pi, pi] Course over ground setpoint, w.r.t. North. NAN if not controlled directly.
+float32 airspeed_direction    		# [rad] [-pi, pi] Angle projected to ground of desired airspeed vector, w.r.t. North. NAN if not controlled directly, used as feedforward if course setpoint is finite.
```
> I think we need to describe this better, as it is not a standard term. Also this does not necessarily have something to do with ground.
Desired angle of the resulting airspeed vector w.r.t. (true) north. Same as vehicle heading iff in the absence of sideslip.

addressed

sfuhrer replied:
>Also can we have a feedforard term here? How can you set a course and still add a feedforward term in heading? This should be either course or airspeed direction, no?

This is still pendnig.

sfuhrer replied:
Removed it in 95883c8d3ca2b9514cc37d97fdab50924648032d

sfuhrer commented on msg/LateralControlLimits.msg:
```diff
@@ -0,0 +1,3 @@
+uint64 timestamp
+
+float32 lateral_accel_max # maps 1:1 to a maximum roll angle, accel_max = tan(roll_max) * GRAVITY, m/s^2
```
I find acceleration the better interface than roll. Also thinking about skid-to-turn at some point.

sfuhrer commented on msg/FwLongitudinalControlSetpoint.msg:
```diff
@@ -0,0 +1,11 @@
+float32 altitude_setpoint  # NAN if not controlled, MSL [m]
+float32 equivalent_airspeed_setpoint # [m/s]
+float32 pitch_sp # NAN if not controlled, overrides total energy controller [rad]
+float32 thrust_sp # NAN if not controlled, overrides total energy controller [0,1]
```
That would require the inverse approach as we now have (currently: flap setting leads to limiting airspeed).
We keep it as is for now.

sfuhrer commented on src/lib/npfg/CourseToAirspeedRefMapper.hpp:
```diff
@@ -0,0 +1,138 @@
+
+	~CourseToAirspeedRefMapper() = default;
+
+	float mapCourseSetpointToHeadingSetpoint(const float bearing_setpoint,
```
We still do airspeed adaption, but to AIRSPD_MAX.

sfuhrer replied:
Agree on that some kind of feedback for "not reachable course" would be good. I created a follow-up task.

sfuhrer commented on src/lib/npfg/DirectionalGuidance.cpp:
```diff
@@ -0,0 +1,319 @@
+	_feas = bearingFeasibility(wind_cross_bearing, wind_dot_bearing, airspeed, wind_speed);
+
+	// we consider feasibility of both the current bearing as well as that on the track at the current closest point
+	const float feas_combined = _feas * _feas_on_track;
```
Let's keep as is for now, created follow up task.

sfuhrer commented on msg/versioned/LateralControlConfiguration.msg:
```diff
@@ -0,0 +1,8 @@
+
+uint64 timestamp          # time since system start (microseconds)
+
+float32 lateral_accel_max # [m/s^2] maps 1:1 to a maximum roll angle, accel_max = tan(roll_max) * GRAVITY
```
I've adapted the wording slightly.

sfuhrer replied:
It doesn't need to be mapped to roll for vehicles that can skid-to-turn (once we have the functionality in PX4)

sfuhrer commented on msg/versioned/LongitudinalControlConfiguration.msg:
```diff
@@ -0,0 +1,17 @@
+float32 throttle_max 			# [0,1]
+float32 climb_rate_target 		# [m/s] target climbrate used to change altitude. Not used if height_rate is directly set in FixedWingLongitudinalSetpoint.
+float32 sink_rate_target 		# [m/s] target sinkrate used to change altitude. Not used if height_rate is directly set in FixedWingLongitudinalSetpoint.
+float32 speed_weight 			# [0,2], 0=pitch controls altitude only, 2=pitch controls airspeed only
+bool enforce_low_height_condition 	# if true, total energy controller will use lower altitude control time constant
+bool disable_underspeed_protection 	# if true, underspeed handling is disabled in the total energy controller
```
I'm not strongly opinionated but rather in favor of keeping it continuously, not just 3 options.

sfuhrer replied:
>Do you want to use a separate message for this

That I don't find a good idea, better have all configuration options in one message.

sfuhrer commented on msg/FixedWingLateralStatus.msg:
```diff
@@ -0,0 +1,7 @@
+# Fixed Wing Lateral Status message
```
The LateralGuidanceStatus message will be combined with this one here once we have all in the same module (planned for with path following).
We could rename it to LateralControllerStatus or ControllerLateralStatus.

sfuhrer commented on src/modules/uxrce_dds_client/dds_topics.yaml:
```diff
@@ -163,5 +163,17 @@ subscriptions:
   - topic: /fmu/in/aux_global_position
     type: px4_msgs::msg::VehicleGlobalPosition
 
+  - topic: /fmu/in/fixed_wing_longitudinal_setpoint
```
I would do the other way around: as the requirements for the FW lateral controller have not changed (wind is as important as before) I reverted moving it to the versioned messages, let's do this later.

sfuhrer commented on src/lib/npfg/NpfgTest.cpp:
```diff
@@ -0,0 +1,398 @@
+
+using namespace matrix;
+
+TEST(NpfgTest, Test)
```
removed it.

sfuhrer commented on src/lib/npfg/NpfgTest.cpp:
```diff
@@ -0,0 +1,398 @@
+
+
+	// THEN: expect heading due South with a min airspeed equal to min_ground_speed
+	EXPECT_NEAR(heading_setpoint, -M_PI_F,  0.01f);
```
Good point, have addressed it.

sfuhrer replied:
9182e4bce76f71613cb3d4dbd5c99a1c7d6b2f3c

sfuhrer commented on posix-configs/SITL/init/test/test_shutdown:
```diff
@@ -29,7 +29,7 @@ flight_mode_manager start
 mc_att_control start vtol
 mc_rate_control start vtol
-fw_pos_control start vtol
+fw_mode_manager start
```
changed that, now always also start fw_lat_lon_control when fw_mode_manager is started.

sfuhrer replied:
01cddd3c67d47fc1ce62f52b8976dba13f236c67

sfuhrer commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.hpp:
```diff
@@ -0,0 +1,264 @@
+#include <uORB/topics/vehicle_status.h>
+#include <uORB/topics/wind.h>
+
+const fixed_wing_lateral_setpoint_s empty_lateral_control_setpoint = {.timestamp = 0, .course = NAN, .airspeed_direction = NAN, .lateral_acceleration = NAN};
+const fixed_wing_longitudinal_setpoint_s empty_longitudinal_control_setpoint = {.timestamp = 0, .altitude = NAN, .height_rate = NAN, .equivalent_airspeed = NAN, .pitch_direct = NAN, .throttle_direct = NAN};
```
done

sfuhrer commented on src/lib/npfg/AirspeedReferenceController.hpp:
```diff
@@ -0,0 +1,75 @@
+#ifndef PX4_AIRSPEEDREFERENCECONTROLLER_HPP
+#define PX4_AIRSPEEDREFERENCECONTROLLER_HPP
+
+class AirspeedReferenceController
```
7d5d5b8c22ac25cea0030dea4369e2e218f58069

sfuhrer commented on src/modules/fw_lateral_longitudinal_control/FwLateralLongitudinalControl.cpp:
```diff
@@ -0,0 +1,841 @@
+	_tecs.set_airspeed_rate_measurement_std_dev(_param_speed_rate_standard_dev.get());
+	_tecs.set_airspeed_filter_process_std_dev(_param_process_noise_standard_dev.get());
+
+	_performance_model.runSanityChecks();
+
+	_performance_model.updateParameters();
```
Makes more sense to switch the order around: 31a496e41d78003bd460d1176d94fb9649f56ee5

sfuhrer commented:
@RomanBapst @mahima-yoga @KonradRudin I've cleaned up the commits (incl. incorporating the changes in https://github.com/PX4/PX4-Autopilot/pull/24886) and force pushed.

mahima-yoga commented on src/lib/npfg/NpfgTest.cpp:
```diff
@@ -0,0 +1,368 @@
+ * Run this test only using "make tests TESTFILTER=Npfg"
+ *
+ *
+ * NOTE:
```
remove?

sfuhrer replied:
well ookay :sweat_smile: If I get other comments on things to change I remove it, otherwise I think we can live with it:)

sfuhrer commented:
I'm going ahead and merge this, after giving this a lot of testing in the last months. Wrapping up the final documentation items in separate PRs. 
Great work everybody!
